lib/src/client/transport/web_transport.dart

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2017, the gRPC project authors. Please see the AUTHORS file
+// for details. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+import 'dart:html';
+import 'dart:typed_data';
+
+import '../../shared/streams.dart';
+
+import '../options.dart';
+
+import 'transport.dart';
+
+const int readyStateDone = 4;
+
+class GrpcWebTransportStream extends GrpcTransportStream {
+  HttpRequest _request;
+  StreamController<ByteBuffer> _incomingProcessor;
+  StreamController<GrpcMessage> _incomingMessages;
+  StreamController<List<int>> _outgoingMessages;
+
+  // TODO: implement id
+  @override
+  int get id => null;
+
+  // TODO: implement incomingMessages
+  @override
+  Stream<GrpcMessage> get incomingMessages => _incomingMessages.stream;
+
+  // TODO: implement outgoingMessages
+  @override
+  StreamSink<List<int>> get outgoingMessages => _outgoingMessages.sink;
+
+  GrpcWebTransportStream(this._request) {
+    _incomingProcessor = new StreamController();
+    _incomingMessages = new StreamController();
+    _outgoingMessages = new StreamController();
+
+    _incomingProcessor.stream
+      .transform(new GrpcWebDecoder())
+      .transform(grpcDecompressor())
+      .listen(_incomingMessages.add,
+        onError: _incomingMessages.addError,
+        onDone: _incomingMessages.close);
+
+    _outgoingMessages.stream
+      .map(GrpcHttpEncoder.frame)
+      .listen((data) {
+        //final encoded = base64Encode(data);
+        _request.send(data);
```

Comment:
One thing that's not clear from wrapping the xhr HttpRequest like this: what's the behavior of client / server / bi-directional streaming cases through this transport type? It looks like in client / directional cases, just send the first request?

Reply:
This gets at some of the other comments about the missing implementations of the other calls.

I might need to find a way to determine whether we're making a single call or a request for bi-directional streaming which would require WebSocket support, not an XHR call. At the moment, the XHR should support the single RPC call and a server stream (although it's buggy at the moment) but not client stream or bi-directional stream.

Apparently client / bi-directional streams are in some gRPC-Web implementaitons, so I just have to figure out how to port them.